lib/Transforms/LiquidFixpointBuilder/VariablesEnvironmentImmutable.cpp

```diff
@@ -0,0 +1,577 @@
+    }
+
+    // Make an assignment to the variable
+    {
+      std::string constraintName = "Variable_"s + mappedVariableName + "_Assignment"s;
+      auto addConstRes = constraintBuilder.AddConstraintForAssignment(constraintName, mappedVariableName, expression, currVariablesAndInfo);
+      if (!addConstRes.Succeeded) { return addConstRes; }
+    }
+
+    // Make updates to Variable Mappings and Values for State Update of the Block
+    insertOrAssignVarType(variableTypes, variable, type);
+    variablesMappingsPerBlock[currentBlockName][variable] = mappedVariableName;
+    variablesValuesPerBlock[currentBlockName].emplace(mappedVariableName);
+
+    return ResultType::Success();
+  }
+  
+  // Create an Immutable Variable
+  ResultType VariablesEnvironmentImmutable::CreateImmutableVariable(
+    std::string variable,
+    FixpointType type,
+    std::vector<std::string> constraints,
+    std::string expression)
+  {
+    if (RefinementUtils::ContainsKey(variableTypes, variable))
+    {
+      return ResultType::Error("Variable"s + variable + " already exists."s);
+    }
+
+    return createVariable(variable, variable, type, constraints, expression);
+  }
+
+  ResultType VariablesEnvironmentImmutable::AddJumpInformation(const std::string& targetBlock)
+  {
+    std::string transitionGuardName = "__transition__"s + currentBlockName + "__"s + targetBlock;
+    {
+      auto createBinderRes = constraintBuilder.CreateBinderWithConstraints(transitionGuardName, FixpointType::GetBoolType(), { "true"s });
+      if (!createBinderRes.Succeeded) { return createBinderRes; }
+    }
+
+    return ResultType::Success();
+  }
+
+  ResultType VariablesEnvironmentImmutable::AddBranchInformation(const std::string& booleanVariable, const bool variableValue, const std::string& targetBlock)
+  {
+    // Verify the variable doesn't already exist
+    if (!RefinementUtils::ContainsKey(variableTypes, booleanVariable))
+    {
+      return ResultType::Error("Missing variable: "s + booleanVariable);
+    }
+
+    // Enforce the variable type is Boolean
+    if (variableTypes.at(booleanVariable) != FixpointType::GetBoolType())
+    {
+      return ResultType::Error("Expected boolean type for variable: "s + booleanVariable);
+    }
+
+    // Convert the variable into a Proposition for a Fixpoint Constraint
+    const std::string assignedExpr = "__value <=> "s + (variableValue ? ""s : "~"s) + booleanVariable;
+    const std::string transitionGuardName = "__transition__"s + currentBlockName + "__"s + targetBlock;
+    
+    {
+      auto createdBinderRes = constraintBuilder.CreateBinderWithConstraints(transitionGuardName, FixpointType::GetBoolType(),  { assignedExpr });
+      if (!createdBinderRes.Succeeded) { return createdBinderRes; }
+    }
+
+    return ResultType::Success();
+  }
+
+  ResultType VariablesEnvironmentImmutable::validateFinishedAndUnfinishedBlockStates(
+    const std::vector<std::string>& previousFinishedBlocks,
```

Comment:
Probably remove this function

Reply:
The upcoming commit may not immediately remove this function. I am first removing redundant variables, and making sure the new logic is not broken because of removals. This will come up next.